src/com/hms_networks/americas/sc/realtimedata/RealTimeDataQueueManager.java

```diff
@@ -0,0 +1,316 @@
+   */
+  public static void dataFetcherB() {
+    for (int i = 0; i < tagManagerB.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupB.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerB.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a new data point for each tag in group C and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherC() {
+    for (int i = 0; i < tagManagerC.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupC.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerC.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a new data point for each tag in group D and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherD() {
+    for (int i = 0; i < tagManagerD.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupD.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerD.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a tag's current value into a list of tag current values
+   *
+   * @param tag the tag to inspect
+   * @return returns a DataPoint object holding the tags current value
+   */
+  private static DataPoint getCurentTagValue(TagInfo tag) {
+
+    String tagName = tag.getName();
+    TagType tagType = tag.getType();
+    int tagID = tag.getId();
+    String timeStampSeconds = String.valueOf(System.currentTimeMillis() / 1000);
+    DataPoint data = null;
+    TagControl tagControl = null;
+
+    try {
+      tagControl = new TagControl(tagName);
+    } catch (EWException e) {
+      e.printStackTrace();
+    }
+
+    if (tagType == TagType.FLOAT) {
+      float val = (float) tagControl.getTagValueAsDouble();
+      data = new DataPointFloat(tagName, tagID, val, timeStampSeconds);
+    } else if (tagType == TagType.INTEGER) {
+      int val = tagControl.getTagValueAsInt();
+      data = new DataPointFloat(tagName, tagID, val, timeStampSeconds);
+    } else if (tagType == TagType.STRING) {
+      String val = tagControl.getTagValueAsString();
+      data = new DataPointString(tagName, tagID, val, timeStampSeconds);
+    } else if (tagType == TagType.BOOLEAN) {
+      boolean val = (tagControl.getTagValueAsLong() != 0);
+      data = new DataPointBoolean(tagName, tagID, val, timeStampSeconds);
+    } else if (tagType == TagType.DWORD) {
+      long val = tagControl.getTagValueAsLong();
+      data = new DataPointDword(tagName, tagID, val, timeStampSeconds);
+    }

```

Comment:
What if the tag control creation failed on 292?

Reply:
added warning logs for this situation